base/shell/explorer/trayntfy.cpp

```diff
@@ -33,6 +34,255 @@ typedef struct _SYS_PAGER_COPY_DATA
+                delete Icon;
+            }
+        }
+        m_WatcherList.RemoveAll();
+
+        LeaveCriticalSection(&m_ListLock);
+    }
+
+    bool AddIconToWatcher(_In_ NOTIFYICONDATA *iconData)
+    {
+        IconWatcherData *Icon = new IconWatcherData(iconData);
+
+        (void)GetWindowThreadProcessId(iconData->hWnd, &Icon->ProcessId);
+
+        Icon->hProcess = OpenProcess(SYNCHRONIZE, FALSE, Icon->ProcessId);
+        if (Icon->hProcess == NULL)
+            return false;
+
+        EnterCriticalSection(&m_ListLock);
+
+        m_WatcherList.AddTail(Icon);
+        SetEvent(m_WakeUpEvent);
+
+        LeaveCriticalSection(&m_ListLock);
+
+        return true;
+    }
+
+    bool RemoveIconFromWatcher(_In_ NOTIFYICONDATA *iconData)
+    {
+        EnterCriticalSection(&m_ListLock);
+        
+        IconWatcherData *Icon;
+        Icon = GetListEntry(iconData, NULL, true);
+
+        SetEvent(m_WakeUpEvent);
+        LeaveCriticalSection(&m_ListLock);
+
+        delete Icon;
+        return true;
+    }
+
+    IconWatcherData* GetListEntry(_In_opt_ NOTIFYICONDATA *iconData, _In_opt_ HANDLE hProcess, _In_ bool Remove)
+    {
+        IconWatcherData *Entry = NULL;
+        POSITION NextPosition = m_WatcherList.GetHeadPosition();
+        POSITION Position;
+        do
+        {
+            Position = NextPosition;
+
+            Entry = m_WatcherList.GetNext(NextPosition);
+            if (Entry)
+            {
+                if ((iconData && ((Entry->IconData.hWnd == iconData->hWnd) && (Entry->IconData.uID == iconData->uID))) ||
+                     (hProcess && (Entry->hProcess == hProcess)))
+                {
+                    if (Remove)
+                        m_WatcherList.RemoveAt(Position);
+                    break;
+                }
+            }
+            Entry = NULL;
+
+        } while (NextPosition != NULL);
+
+        return Entry;
+    }
+
+private:
+
+    static UINT WINAPI WatcherThread(_In_opt_ LPVOID lpParam)
+    {
+        CIconWatcher* This = reinterpret_cast<CIconWatcher *>(lpParam);
+
+        This->m_Loop = true;
+        while (This->m_Loop)
+        {
+            HANDLE *WatchList;
+            DWORD Size;
+
+            EnterCriticalSection(&This->m_ListLock);
+
+            Size = This->m_WatcherList.GetCount() + 1;
+            WatchList = new HANDLE[Size];
+            WatchList[0] = This->m_WakeUpEvent;
+
+            POSITION Pos;
+            for (size_t i = 0; i < This->m_WatcherList.GetCount(); i++)
+            {
+                Pos = This->m_WatcherList.FindIndex(i);
+                if (Pos)
+                {
+                    IconWatcherData *Icon;
+                    Icon = This->m_WatcherList.GetAt(Pos);
+                    WatchList[i + 1] = Icon->hProcess;
+                }
+            }
+
+            LeaveCriticalSection(&This->m_ListLock);
+
+            DWORD Status;
+            Status = WaitForMultipleObjects(Size,
```

Comment:
Your thread will die if there are more than 64 processes with tray icons. Might want to handle that more gracefully.

Reply:
Yep, that's a good point, I'd overlooked that.
Thanks Thomas.

Reply:
Also, the thread should close all the handles (and cleanup the list) if it dies prematurely, otherwise we risk zombie processes